src/checkpoint/traits/user_traits.h

```diff
@@ -0,0 +1,196 @@
+//
+// Copyright 2019 National Technology & Engineering Solutions of Sandia, LLC
+// (NTESS). Under the terms of Contract DE-NA0003525 with NTESS, the U.S.
+// Government retains certain rights in this software.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are met:
+//
+// * Redistributions of source code must retain the above copyright notice,
+//   this list of conditions and the following disclaimer.
+//
+// * Redistributions in binary form must reproduce the above copyright notice,
+//   this list of conditions and the following disclaimer in the documentation
+//   and/or other materials provided with the distribution.
+//
+// * Neither the name of the copyright holder nor the names of its
+//   contributors may be used to endorse or promote products derived from this
+//   software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact darma@sandia.gov
+//
+// *****************************************************************************
+//@HEADER
+*/
+
+#if !defined INCLUDED_SRC_CHECKPOINT_TRAITS_USER_TRAITS_H
+#define INCLUDED_SRC_CHECKPOINT_TRAITS_USER_TRAITS_H
+
+namespace {
+  struct NoTrait;
+
+  template<typename Traits, typename T, typename... U>
+  struct without_helper {
+    using type = typename without_helper<typename Traits::template _without_trait<T>, U...>::type;
+  };
+  template<typename Traits, typename T>
+  struct without_helper<Traits, T> {
+    using type = typename Traits::template _without_trait<T>;
+  };
+}
+
+namespace checkpoint {
+namespace SerializerUserTraits {
+
+  template<typename U, typename V>
+  struct is_base_or_derived
+    : std::disjunction<std::is_base_of<U,V>, std::is_base_of<V,U>> {};
+
+  template<typename U, typename V>
+  struct is_same_template
+    : std::false_type {};
+  template<template <typename...> typename Templ, typename... U, typename... V>
+  struct is_same_template<Templ<U...>, Templ<V...>>
+    : std::true_type {};
+
+  template<typename U, typename V>
+  struct traits_match
+    : std::disjunction<
+#ifdef CHECKPOINT_INHERITED_USER_TRAITS
```

Comment:
They aren't - I wasn't sure if it would be best to have these extra checks or not, so I figured I'd defer the question with these. Think I should drop the ifdefs or drop the checks?